### PR TITLE
Boost modification for IBM xl compiler

### PR DESCRIPTION
**Pull request by eklee15**
Added patch for IBM xl (xl_1_62_0_le.patch) for boost 1.62.0
Modified /boost/package.py

### CODE REVIEW

File: var/spack/repos/builtin/packages/boost/package.py [hunk 1]
```diff
@@ -126,13 +126,20 @@ class Boost(Package):
     depends_on('mpi', when='+mpi')
     depends_on('bzip2', when='+iostreams')
     depends_on('zlib', when='+iostreams')
+    depends_on('zlib', when='%xl_r' )
```
Comment (davydden) [inline var/spack/repos/builtin/packages/boost/package.py]: it's not obvious why the two dependencies are needed for xl compiler regardless of `iostreams` variant.

File: var/spack/repos/builtin/packages/boost/package.py [hunk 2]
```diff
@@ -126,13 +126,20 @@ class Boost(Package):
...
+    depends_on('bzip2', when='%xl_r')
 
     # Patch fix from https://svn.boost.org/trac/boost/ticket/11856
     patch('boost_11856.patch', when='@1.60.0%gcc@4.4.7')
 
     # Patch fix from https://svn.boost.org/trac/boost/ticket/10125
     patch('boost_10125.patch', when='@1.55.0%gcc@5.0:5.9')
 
+    # Patch fix for IBM XL compiler
+    patch('xl_1_62_0_le.patch', when='@1.62.0%xl_r')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/boost/package.py]: is the patch submitted upstream?

File: var/spack/repos/builtin/packages/boost/package.py (hunk 1 above)
Comment (eklee15) [inline var/spack/repos/builtin/packages/boost/package.py]: I would like to build zlib and bzip2 dependencies with gcc, but was not able to do it without it. 

spack install boost@1.62.0%xl_r^zlib%gcc^bzip2%gcc

File: var/spack/repos/builtin/packages/boost/package.py (hunk 2 above)
Comment (eklee15) [inline var/spack/repos/builtin/packages/boost/package.py]: yes.

File: var/spack/repos/builtin/packages/boost/package.py (hunk 1 above)
Comment (davydden) [inline var/spack/repos/builtin/packages/boost/package.py]: well, then this most certainly should NOT be a part of this PR. Please open a separate issue to discuss limitations of conretizer.

Comment (davydden) [inline var/spack/repos/builtin/packages/boost/package.py]:
```
spack install boost@1.62.0+iostreams%xl_r^zlib%gcc^bzip2%gcc
```
would probably work.

Comment (eklee15) [inline var/spack/repos/builtin/packages/boost/package.py]: Make sense, I will modify it. Thanks.

Comment (davydden) [inline var/spack/repos/builtin/packages/boost/package.py]: At the moment, those constrains are easier to specify in `packages.yaml` as opposed to CLI. http://spack.readthedocs.io/en/latest/configuration.html

**Comment by adamjstewart** [comment]
Can you fix the flake8 warnings?:
```
var/spack/repos/builtin/packages/boost/package.py:141: [E303] too many blank lines (2)
var/spack/repos/builtin/packages/boost/package.py:158: [E231] missing whitespace after ':'
var/spack/repos/builtin/packages/boost/package.py:159: [E127] continuation line over-indented for visual indent
var/spack/repos/builtin/packages/boost/package.py:159: [E231] missing whitespace after ':'
```

File: var/spack/repos/builtin/packages/boost/package.py (hunk 2 above)
Comment (tgamblin) [inline var/spack/repos/builtin/packages/boost/package.py]: Is this the only version we can build with XL? Or does the patch apply just as well to versions other than 1.62.0? And what about regular `xl`?

Comment (eklee15) [inline var/spack/repos/builtin/packages/boost/package.py]: Should work with xl, but have not tested it. I'm planning to make another PR after testing with xl and other versions of boost (i.e., 1.63.0).
